### ISSUE TITLE
Create transparent window on top of everything

### ISSUE THREAD

**Issue by evgenyzinoviev** [atom/electron#3888]
Is it possible to create a full-screen transparent window on top of all running applications, dock, menubar (on OSX) an everything else with BrowserWindow or I have to hack around with Electron sources?

I believe this has something to do with window level (https://developer.apple.com/library/mac/documentation/Cocoa/Reference/ApplicationKit/Classes/NSWindow_Class/#//apple_ref/doc/constant_group/Window_Levels), but I could not make what I want with BrowserWindow, so I'm asking here.

**Comment by zcbenz** [comment]
No, currently we don't have API for that.

**Comment by evgenyzinoviev** [comment]
It would be nice to have!

**Comment by zcbenz** [comment]
After some research I don't think there is actually a way to achieve this, but I might be wrong since I'm not an expert of Cocoa. Anyway I'm closing this since I don't think it is possible, but if anyone came up with a solution I would be happy to accept a pull request.

**Comment by evgenyzinoviev** [comment]
Ok I will dig into this soon.

**Comment by evgenyzinoviev** [comment]
Just figured out how to achieve this. The key thing is to set window level ```[window_ setLevel:NSScreenSaverWindowLevel];```, cocoa_bounds to size of screen ```cocoa_bounds = [NSScreen mainScreen].frame;``` and ```setEnableLargerThanScreen``` to ```true```. Also blocking of moving and resizing is recommended.

I'll think of the best way to add this to Electron and send you a pull request.

**Comment by evgenyzinoviev** [comment]
@zcbenz Could you (or someone) please help me. I've already done this for OS X and now I'm trying to achieve this on Windows. After that I'll send a PR. But there's a little problem. 

The size of my screen is 1440x900. I'm creating a BrowserWindow with params:
```javascript
{ x: 0, y: 0, width: 1440, height: 900 }
```

To make the window topmost, I call `::SetWindowPos(GetAcceleratedWidget(), HWND_TOPMOST, x, y, width, height, 0)`, it's working fine.

On these lines (`native_window_views.cc:127`) the `widget_size_` is 1440x900 as expected:
```c++
gfx::Rect bounds(0, 0, width, height);
widget_size_ = bounds.size();
```
But then `NativeWindowViews::OnWidgetBoundsChanged` fires and sets height to `860`. I guess that 40px that is gone is a taskbar. As a result, the window not overlapping the taskbar.

But if I do a dirty hack and reset window size like this:
```c++
NativeWindow::SetSize(bounds.size())
```
in the end of `NativeWindowViews` constructor, then the window overlaps everything.

So basically the problem is that something somewhere shrinks the bounds after the window is created, I'm not an expert in Electron's code, do you know what is it and how to prevent it doing this?

**Comment by zcbenz** [comment]
@evgenyzinoviev Chromium is maintaining a internal state of window size, so if you call win32 API to change the window size Chromium might revert it back later. You can probably just call `SetAlwaysOnTop` to make the window top most.

**Comment by evgenyzinoviev** [comment]
@zcbenz 
1. Ok, I will not touch window size on Windows.
2. Yes, `alwaysOnTop` seems to work on Windows.
But not on OS X. It sets `windowLevel` to `NSFloatingWindowLevel` and this is what it looks like: 
<img src="https://cloud.githubusercontent.com/assets/6204306/12537582/bc092206-c2c2-11e5-9513-f052d2b784ba.png" width="400" height="auto">

And this what I need (`NSScreenSaverWindowLevel`):
<img src="https://cloud.githubusercontent.com/assets/6204306/12537590/e7f4bc90-c2c2-11e5-890a-a9dbc255e59d.png" width="400" height="auto">

I think we shouldn't break how setAlwaysOnTop works now, but we should either add a new option for OS X (like `topMost: true`) or new window `type` (like `topmost` or `overlay`). Which is better? I'll send a PR.

**Comment by zcbenz** [comment]
A new window type for OS X sounds good to me, `topMost` looks very similar to `alwaysOnTop` so it is going to cause confusions.

**Comment by evgenyzinoviev** [comment]
Hmm, I've started and now I think that we are mixing slightly different kinds of things here: types and levels. I might want (and I actually need to) separately change window level and textured background on OS X (as in `textured` window type).

How about making option `level` with supported values like `desktop`, `normal`, `floating`, `screensaver` (or any other name for topmost level, probably even `topmost`) and methods `setLevel`, `getLevel`.

The `type` option will remain only for `textured` on OS X and without changes for Linux (`textured` is not actually a type, it sets window background mask, but ok).

`alwaysOnTop` option will stay for backward compatibility.

**Comment by zcbenz** [comment]
Adding new `level` option sounds good to me, we can make it OS X only.

**Comment by matthieualouis** [comment]
Hi, 

I am trying to build an electron app that behaves in the same way as the mac Helium App http://jadengeller.github.io/Helium/  (i.e. it stays on top of other windows even when not focused). 

Would it be possible to achieve this behaviour with the your suggestion @evgenyzinoviev? Or do I have to get my hands dirty with the accessibility API.

Thanks!

**Comment by evgenyzinoviev** [comment]
@matthieualouis  Yes, it's possible. I can make a pull request with this feature soon.

**Comment by matthieualouis** [comment]
That would be awesome :) , let me know if there is anything I can do to help!

**Comment by ericsaboia** [comment]
that will be really useful

**Comment by matthieualouis** [comment]
Hey @evgenyzinoviev , do you have any update on this feature?

**Comment by bobasoft** [comment]
@evgenyzinoviev any update on window level option for OS X? Really need the solution, have tried to set window level via getNativeWindowHandle() and NodObjC, but couldn't make it work.

**Comment by matheuss** [comment]
@evgenyzinoviev any update on this? It will be very useful 😄

**Comment by CodeTheory** [comment]
Is anyone else able to handle this? This would be a brilliant addition.

**Comment by kevinsawicki** [comment]
I believe this may have been addressed by #7296 via the`level` option to http://electron.atom.io/docs/api/browser-window/#winsetalwaysontopflag-level

**Comment by CodeTheory** [comment]
@kevinsawicki I'll experiment with this tomorrow. Thanks!

**Comment by robsontenorio** [comment]
`BrowserWindow.setAlwaysOnTop` with any `level`does not allow the electron app stay on top fullscreen apps (OSX).
